File: 프로그래머스/unrated/160586. 대충 만든 자판/대충 만든 자판.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int[] solution(String[] keymap, String[] targets) {
        int[] answer = new int[targets.length];
        Map<String, Integer> map = new HashMap<>();

        for (int i = 0; i < keymap.length; i++) {
            for (int j = 0; j < keymap[i].length(); j++) {
                String c = String.valueOf(keymap[i].charAt(j));

                if (!map.containsKey(c) || map.get(c) > j+1){
                    map.put(c, j + 1);
                }
            }
        }

        for (int i = 0; i < targets.length; i++) {
            for (int j = 0; j < targets[i].length(); j++) {
                String targetC = String.valueOf(targets[i].charAt(j));
                if (map.containsKey(targetC)){
                    answer[i] += map.get(targetC);
                } else {
                    answer[i] = -1;
                    break;
                }

            }
        }

        return answer;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        String[] keymap = {"ABACD", "BCEFD"};
        String[] targets = {"ABCD","AABB"};
        int[] answer = sol.solution(keymap, targets);
        System.out.println("answer = " + Arrays.toString(answer));
    }
}

File: JavaScript/프로그래머스/1/160586. 대충 만든 자판/대충 만든 자판.js
function solution(keymap, targets) {
    let answer = [];
    let map = {};

    for (let key of keymap) {
        for (let i = 0; i < key.length; i++) {
            if (map[key[i]] === undefined || i < map[key[i]]) {
                map[key[i]] = i;
            }
        }
    }
    
    for (target of targets) {
        let count = 0;
        for (let i = 0; i < target.length; i++) {
            if (map[target[i]] !== undefined) {
                count += map[target[i]] + 1;
            } else {
                count = -1;
                break;
            }
        }
        answer.push(count);
    }

    return answer;
}